nav2_regulated_pure_pursuit_controller/include/nav2_regulated_pure_pursuit_controller/path_handler.hpp

```diff
@@ -76,10 +76,19 @@ class PathHandler
     const geometry_msgs::msg::PoseStamped & in_pose,
     geometry_msgs::msg::PoseStamped & out_pose) const;
 
-  void setPlan(const nav_msgs::msg::Path & path) {global_plan_ = path;}
+  void setPlan(const nav_msgs::msg::Path & path)
+  {
+    global_plan_ = path;
+    path_progress_cursor_ = global_plan_.poses.begin();
```

Comment:
This seems to accomplish the same thing as the path erasing logic did previously without storing extra information (and erasing that which is no longer relevant). Can you explain this change?

Reply:
Yes, this is designed to do the same thing as the path erasing logic. However, the key difference is retaining the erased poses so that the cross-track error calculation can look backward to those if it needs to. Otherwise, the cross-track error is too spiky since you are always off the end of the pruned path.

As for the pruning/erasing/cursor logic in general, it remains relevant since it is half of what keeps the closest waypoint logic from snapping to waypoints that have already been traveled past.

Reply:
Is there a reason not to just do the erase but just keep `i-1`'s value? I think the cross track error calculation just uses the `i-1` and `i+1` th values

Reply:
Yes that would work, but what do you gain by erasing vs just keeping a cursor? It's probably negligible, but erasing probably uses more computation (requires reallocating the path every time).

Reply:
Well, maybe that part about reallocation is not true, but it will relocate all elements after the erased portion: https://en.cppreference.com/w/cpp/container/vector/erase
> Complexity
Linear: the number of calls to the destructor of T is the same as the number of elements erased, the assignment operator of T is called the number of times equal to the number of elements in the vector after the erased elements

Reply:
Its one less thing being stored as state that can get screwed up in the future :-) 

Or someone that didn't pay attention closely not realizing they needed to start at the cursor would just start at `begin()` because its seems reasonable for a new feature. 

Mostly, just simpler bookkeeping

Reply:
That makes sense that it could be a point of mistakes where you accidentally use begin. Although I would argue that explicit state is better than implicit state.

Also, if you have to store one pose backwards, then you have the same problem where you could use `begin()` accidentally erroneously, since most of the references will have to change to `begin() + 1`.

Reply:
I think this is just a point of preference. In general, if it were just you and I modifying this code, I think that I wouldn't care very much. However, that `begin()` versus using the cursor is a subtle mistake that wouldn't be immediately clear to me when reviewing a PR if it wasn't loaded in my brain's RAM that day and I could see that introducing a bug because of the nature of open-source. Since all the other instances of things like this operate in this way, this would be a single special case. If that was absolutely necessary, then I wouldn't argue with it. 

But if there's a way to keep that consistency, its less likely that 3 years from now we introduce a sutble bug when this is long since out of thought / mind.

> Also, if you have to store one pose backwards, then you have the same problem where you could use begin() accidentally erroneously, since most of the references will have to change to begin() + 1.

I don't understand. We don't assume that the first point is the closest necessarily. We're just assuming that we've removed absolutely erroneous information.

Reply:
No what I'm saying is that if you remove the cursor, and use the existing erase logic but back pad one pose, then you have to convert all other references of begin() to begin() + 1 to compensate. And it's very possible that we need to back pad more than one pose, so then you should probably store that constant offset in a variable, like pose_look behind or something. And then all the references that iterate through that path (except path error) need to change to begin() + pose_lookbehind. And my point is that just storing an iterator that points to the point in the path you are interested in is more explicit than remembering to use begin and an offset variable. No matter whether you use an iterator cursor or not, there are many places in the code where begin() alone is not correct, and erasing doesn't make that pitfall less likely.

Reply:
So I think the choice is not whether to have a separate variable or not. The choice is whether that variable is an offset (which would likely be parameterized), or an iterator (which would not require a separate parameter for how long the lookbehind should be).

Reply:
> convert all other references of begin() to begin() + 1 to compensate

Why? I agree if we had to do something like that its better to use the cursor. 

I don't see why back padding by 1 path point is necessary in any of the loops. We're never assuming that the first point has any special characteristic (e.g. closest point). We just assume that we've erased any points before that point which are no longer relevant.

Reply:
OK that makes sense, lets keep the cursor. Lets just make sure to update all appropriate locations for it to be used :-) I don't know if there's somewhere natural to document this to make it clear (or rename the path object), but spend a minute or two trying to see if we can prevent future idiots like me from screwing this up :wink: 

The "best" solution, which I'm not necessarily asking you to do, would be to make a new struct containing the path and the cursor with some functions for iterating through it that keeps track of the cursor. That seems to me to be the most idiot-proof solution, but you might have something more clever!